runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/streaming/WeightBoundedQueueTest.java

```diff
@@ -0,0 +1,130 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.streaming;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class WeightBoundedQueueTest {
+  private static final int MAX_WEIGHT = 10;
+
+  @Test
+  public void testPut_hasCapacity() {
+    WeightedBoundedQueue<Integer> queue =
+        WeightedBoundedQueue.create(MAX_WEIGHT, i -> Math.min(MAX_WEIGHT, i));
+
+    int insertedValue = 1;
+
+    queue.put(insertedValue);
+
+    assertEquals(insertedValue, queue.queuedElementsWeight());
+    assertEquals(1, queue.size());
+    assertEquals(insertedValue, (int) queue.poll());
+  }
+
+  @Test
+  public void testPut_noCapacity() throws InterruptedException {
+    WeightedBoundedQueue<Integer> queue =
+        WeightedBoundedQueue.create(MAX_WEIGHT, i -> Math.min(MAX_WEIGHT, i));
+
+    // Insert value that takes all the capacity into the queue.
+    Thread thread1 = new Thread(() -> queue.put(MAX_WEIGHT));
+    thread1.start();
+    thread1.join();
+
+    // Try to insert another value into the queue. This will block since there is no capacity in the
+    // queue.
+    Thread thread2 = new Thread(() -> queue.put(MAX_WEIGHT));
+    thread2.start();
+
+    // Should only see the first value in the queue, since the queue is at capacity.  thread2
+    // should be blocked.
+    assertEquals(MAX_WEIGHT, queue.queuedElementsWeight());
+    assertEquals(1, queue.size());
+
+    // Have another thread poll the queue, pulling off the only value inside and freeing up the
+    // capacity in the queue.
+    Thread thread3 = new Thread(queue::poll);
+    thread3.start();
+    thread3.join();
+
+    // Wait for the thread2 which was previously blocked due to the queue being at capacity.
+    thread2.join();
+
+    assertEquals(MAX_WEIGHT, queue.queuedElementsWeight());
+    assertEquals(1, queue.size());
+  }
+
+  @Test
+  public void testPoll() {
+    WeightedBoundedQueue<Integer> queue =
+        WeightedBoundedQueue.create(MAX_WEIGHT, i -> Math.min(MAX_WEIGHT, i));
+
+    int insertedValue1 = 1;
+    int insertedValue2 = 2;
+
+    queue.put(insertedValue1);
+    queue.put(insertedValue2);
+
+    assertEquals(insertedValue1 + insertedValue2, queue.queuedElementsWeight());
+    assertEquals(2, queue.size());
+    assertEquals(insertedValue1, (int) queue.poll());
+    assertEquals(1, queue.size());
+  }
+
+  @Test
+  public void testPoll_emptyQueue() {
+    WeightedBoundedQueue<Integer> queue =
+        WeightedBoundedQueue.create(MAX_WEIGHT, i -> Math.min(MAX_WEIGHT, i));
+
+    assertNull(queue.poll());
+  }
+
+  @Test
+  public void testTake() throws InterruptedException {
```

Comment:
add a variant where you poll with very large timeout and ensure it gets the element

Reply:
done used polling at 1 minute, inserting element at 30 seconds, so poll should return the element.

Reply:
Thanks, I just meant something like testPoll_withTimeout with a large timeout we never expected to get hit.  You could remove testPoll_withLargeTimeout since I don't think it adds much beyond withTimeout and theses tests don't seem to run with much parallelism and we don't want to slow the build down too much.

Reply:
sg removed!